### DISCUSSION TITLE
Does Next.js support custom entries in the package.json exports field?

### DISCUSSION THREAD

**Discussion by idbartosz** [vercel/next.js#33813] [Help]
**Question**: Does Next.js support custom entries in the package.json exports field?

**Usecase**:
In a monorepo that contains a component package, I've got some custom entries defined for various tools and environments. So far every tool respected the custom entires and everything worked smoothly. But in the case of Next.js I'm not able to force it to use the custom `next` entry, it always tries to read the `"."` or `".": { "import": "..." }` which in my case ends up in an error message `Module not found: Package path . is not exported from package ...`.

**Solutions tried:**
- Adding a `resolve.conditionNames = ['next']` to the weback configuration in `next.config.js` _[1]_
- Adding node options to the development script `"dev": "NODE_OPTIONS='--conditions=next' next dev"` _[2]_
- Trying both at the same time

_[1] https://webpack.js.org/configuration/resolve/#resolveconditionnames_
_[2] https://nodejs.org/api/packages.html#resolving-user-conditions_

**Example**:
_Exports field defined in component package.json_
```json
  "exports": {
    ".": {
      "storybook": "./src/index.ts",
      "next": "./dist/index.js",
    }
  }
```

**Comment by arrizalamin** [comment]
I solved this with a webpack plugin:

```javascript
class NextEntryPlugin {
  apply(compiler) {
    compiler.hooks.afterEnvironment.tap("NextEntryPlugin", () => {
      compiler.options.resolve.conditionNames.unshift("next");
      compiler.options.resolve.mainFields.unshift("next");
    });
  }
}
```

**Reply by jgoux** [reply]
Hey @arrizalamin I'm trying to do the same with Next.js v13.4.2 and it doesn't seem to work. Could you provide the full webpack config example?

**Reply by joostmeijles** [reply]
Appending to `conditionNames` works for me:
```
export function webpack(config) {
  const name = "myConditionName";

  class NextEntryPlugin {
    apply(compiler) {
      compiler.hooks.afterEnvironment.tap("NextEntryPlugin", () => {
        compiler.options.resolve.conditionNames = [...compiler.options.resolve.conditionNames, name]
      });
    }
  }

  config.plugins.push(new NextEntryPlugin());
}
```

**Comment by jgoux** [comment]
@timneutkens Would it be possible to turn this discussion into an issue? I would expect `transpilePackages` to honor `tsconfig.json`.

So if I define: 
```json
"compilerOptions": {
  "customConditions": ["snaplet_development"]
}
```

I would expect Next.js to match this condition when resolving the external packages.

And even better, if the external package is detected as a Typescript project reference, I would also expect Next.js to pick-up the external package's own `tsconfig.json` to resolve its internal path aliases. 😄

Those are the two things I'm missing to make Next.js great in TypeScript monorepos.

**Reply by Kuchasz** [reply]
Hi @jgoux did you found out how to make it to work? If not - have you found any workarounds?
